testplugins/src/main/java/org/spongepowered/test/DamageSourceTest.java

```diff
@@ -0,0 +1,109 @@
+import org.spongepowered.api.event.cause.entity.damage.DamageTypes;
+import org.spongepowered.api.event.cause.entity.damage.source.DamageSource;
+import org.spongepowered.api.event.entity.DamageEntityEvent;
+import org.spongepowered.api.event.filter.Getter;
+import org.spongepowered.api.event.game.state.GameInitializationEvent;
+import org.spongepowered.api.plugin.Plugin;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.format.TextColors;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+@Plugin(id = "damagesourcetest", name = "Damage Source Test", description = "A plugin to test damage sources")
+public final class DamageSourceTest {
+
+    private final Set<UUID> activated = new HashSet<>();
+
+    @Listener
+    public void onInit(GameInitializationEvent event) {
+        Sponge.getCommandManager().register(this,
+                CommandSpec.builder()
+                        .executor((src, args) -> {
+                            if (src instanceof Player) {
+                                final UUID uuid = ((Player) src).getUniqueId();
+                                if (this.activated.contains(uuid)) {
+                                    this.activated.remove(uuid);
+                                    src.sendMessage(Text.of("You have deactivated damage source analysis."));
+                                } else {
+                                    this.activated.add(uuid);
+                                    src.sendMessage(Text.of("You have activated damage source analysis."));
+                                }
+                            }
+                            return CommandResult.success();
+                        })
+                        .build(),
+                "dstest");
+
+        final DamageSource damageSource = DamageSource.builder()
+                .type(DamageTypes.CUSTOM)
+                .exhaustion(5)
+                .scalesWithDifficulty()
+                .build();
+
+        Sponge.getCommandManager().register(this,
+                CommandSpec.builder()
+                        .executor((src, args) -> {
+                            if (src instanceof Player) {
+                                final Player player = (Player) src;
+                                player.damage(args.<Double>getOne("damage").orElse(2.0), damageSource);
+                                player.sendMessage(Text.of("You have damaged yourself with the custom damage source."));
+                            }
+                            return CommandResult.success();
+                        })
+                        .arguments(GenericArguments.doubleNum(Text.of("damage")))
+                        .build(),
+                "dsdamage");
+    }
+
+    @Listener(order = Order.POST)
+    public void onPlayerDamage(DamageEntityEvent event, @Getter("getTargetEntity") Player player) {
+        event.getCause().first(DamageSource.class).ifPresent(d -> {
```

Comment:
Could do with another filter instead of a nested lambda.

Reply:
Addressed.